[BE] refactor: 관리자 페이지 파일 수정시 S3 반영 #246

gabean13 opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->

S3 삭제는 기도메타해야함..
이 부분 맞는지 멍구가 한번 체크해주세요ㅠㅠ

AdminHearitController의 updateHearit ~~ 메서드        
AdminHearitService의 modifyHearitFile 메서드
Hearit 
FileType 

4개 클래스 위주로 봐주세용

FileType 부분 코드가 좀 복잡해지긴했는데 안그러면 Hearit에서 if 분기문이 너무 많아져서 OCP를 너무 안지키는 것 같아서 일단 enum에서 추상 메서드 구현하도록 했습니다.

## 📸 스크린샷 (선택)
> UI 변경이 있을 경우 첨부

메타 데이터 수정 / 파일 수정 버튼 분리
<img width="1859" height="201" alt="image" src="https://github.com/user-attachments/assets/02161f69-d351-4bde-b2be-3dc8b79092ac" />


메타 데이터 수정 시 파일 수정 막음
<img width="1862" height="153" alt="image" src="https://github.com/user-attachments/assets/a97b8901-2e79-44c7-9a04-0754057c4e24" />

## 🧪 테스트 방법 (선택)
<!-- 코드 리뷰하는 팀원이 돌려볼 수 있도록 작성해주기 
ex. ReservationTest의 XXXX 테스트메서드로 확인 
ex. 어떤화면의 어떤기능에서 확인 -->

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->

## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->

close #246


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **신규 기능**
  * 개별적으로 원본 오디오, 요약 오디오, 스크립트 파일을 업데이트할 수 있는 엔드포인트가 추가되었습니다.
  * 파일명 및 파일 일관성 검증 로직이 강화되었습니다.
  * 파일 삭제 기능이 추가되었습니다.

* **개선 사항**
  * 히어릿 메타데이터(제목, 요약, 재생 시간 등)만 별도로 수정할 수 있도록 분리되었습니다.
  * 파일 업로드 및 메타데이터 수정 시 검증이 더욱 엄격해졌습니다.
  * 관리 페이지에서 파일 업로드 UI가 개선되었습니다.

* **버그 수정**
  * 파일명 형식 및 일관성 미준수 시 오류 메시지가 명확하게 표시됩니다.

* **테스트**
  * 파일별 업데이트 및 메타데이터 수정에 대한 테스트가 추가되었습니다.

* **기타**
  * 코드 스타일 및 예외 처리 방식이 일부 변경되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gabean13 committed:
[BE] chore: 파일 업로드 view 수정 #246

gabean13 committed:
[BE] feat: 파일 수정 시 S3 삭제 및 재업로드 기능 구현 #246

- 파일 검증
- 파일 삭제 후 재 업로드

gabean13 committed:
[BE] test: 파일 수정 시 S3 삭제 및 재업로드 테스트 작성 #246

gabean13 committed:
[BE] style: 코드 정렬 #246

HamBeomJoon commented:
좋아요

JO-eusan commented on backend/src/main/java/com/onair/hearit/admin/application/AdminHearitService.java:
```diff
@@ -69,7 +70,10 @@ private Map<Long, List<KeywordInHearit>> mapKeywordsByHearitId(List<HearitKeywor
     @Transactional
     public void addHearit(HearitCreateRequest request) {
-        FileValidator.validateAll(request.originalAudio(), request.shortAudio(), request.scriptFile());
+        FileNameValidator.validateAll(
```
`FileNameValidator`가 꼭 필요할까요?
이 내부에 들어 있는 검증 로직들이 중요하게 생각되는데 이건 url을 가지고 있는 도메인인 `Hearit`에서 검증하는게 맞지 않을까요?

저는 이 함수가 다음과 같은 순서로 되야한다고 생각하는데 가콩의 생각이 궁금합니다.

```java
Category category = getCategoryById(request.categoryId());
Hearit hearit = new Hearit(request.title(), request.summary(), request.playTime(), originalAudioPath, shortAudioPath, scriptFilePath, request.source(), category); // 검증 로직 내부에서 처리
String originalAudioPath = fileStorageService.uploadFile(request.originalAudio(), FileType.ORIGINAL);
String shortAudioPath = fileStorageService.uploadFile(request.shortAudio(), FileType.SHORT);
String scriptFilePath = fileStorageService.uploadFile(request.scriptFile(), FileType.SCRIPT);
        
```

gabean13 replied:
주말에 이슈파서 레벨1의 정신을 이어 객체지향적으로 리팩토링하겠ㅅ브니다.

JO-eusan commented on backend/src/main/java/com/onair/hearit/admin/application/AdminHearitService.java:
```diff
@@ -109,12 +113,24 @@ private void validateHearitKeywords(List<Long> keywordIds, List<Keyword> keyword
     @Transactional
-    public void modifyHearit(Long hearitId, HearitUpdateRequest request) {
+    public void modifyHearitMetaData(Long hearitId, HearitMetaDataUpdateRequest request) {
         Category category = getCategoryById(request.categoryId());
```
`getCatoryById()` 메서드도 `getHearitById()` 메서드 전에 나오도록 순서 변경하면 좋을 것 같아요!

gabean13 replied:
했음둥

JO-eusan commented on backend/src/main/java/com/onair/hearit/admin/application/AdminHearitService.java:
```diff
@@ -109,12 +113,24 @@ private void validateHearitKeywords(List<Long> keywordIds, List<Keyword> keyword
+    @Transactional
+    public void modifyHearitFile(Long hearitId, HearitFileUpdateRequest request, FileType fileType) {
+        Hearit hearit = getHearitById(hearitId);
+        FileNameValidator.validateFileUrl(request.file().getOriginalFilename(), fileType, hearit);
```
이 검증도 `Hearit`에게 메시지를 전달해서 검증하면 될 것 같습니다!
개인적으로 `FileNameValidator`는 검증 책임을 도메인이 아닌 다른 객체에게 위임함으로써 도메인 규칙이 분산되어 있다는 느낌이 있는것 같아요.
가콩은 어덯게 생각하시나요?

yuyoungrhee replied:
이부분 가콩과 잠깐 말했었는데 Hearit 내에 File3개가 Embedded로 VO로 분리하면 좋겠다는 얘기 했었어요.
음.. 관리자 페이지 삭제 급한 거 아니니까 지금 refactor 하는 김에 하면 좋을 거 같다는 생각이 들기도 하고? ㅋㅋ 천천히 합시다..^^ 

이부분은 천천히 반영해도 된다고 생각해서 가콩 의견에 따를게여

JO-eusan commented on backend/src/main/java/com/onair/hearit/admin/application/FileStorageService.java:
```diff
@@ -44,4 +44,15 @@ private String uploadToS3(MultipartFile multipartFile, String key) {
     }
+
+    public void deleteFile(String filePath) {
+        String key = filePath.startsWith("/") ? filePath.substring(1) : filePath;
```
3항 연산자 보다는 다음과 같은 코드는 어떨까요?

```java
String key = filePath;
if(filePath.startWith("/")) {
    key = filePath.substring(1);
}
```

JO-eusan commented on backend/src/main/java/com/onair/hearit/admin/dto/request/HearitFileUpdateRequest.java:
```diff
@@ -0,0 +1,9 @@
+package com.onair.hearit.admin.dto.request;
+
+import org.springframework.web.multipart.MultipartFile;
+import software.amazon.awssdk.annotations.NotNull;
```
단순 궁금증인데 다른 라이브러리의 `@NotNull`을 사용하신 이유가 있나요?

JO-eusan commented on backend/src/main/java/com/onair/hearit/domain/FileType.java:
```diff
@@ -0,0 +1,61 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum FileType {
```
파일을 Enum으로 관리하니까 도메인 규칙도 잘보이고 좋네요!!

JO-eusan commented on backend/src/main/java/com/onair/hearit/domain/FileType.java:
```diff
@@ -0,0 +1,61 @@
+    private final String extension;
+    private final String uploadPath;
+
+    public abstract void updateFileUrl(Hearit hearit, String newUrl);
```
메서드 사이에 한 줄 띄는거 어때용?

gabean13 replied:
했음둥

JO-eusan commented on backend/src/test/java/com/onair/hearit/domain/HearitTest.java:
```diff
@@ -0,0 +1,117 @@
+        // then
+        assertThat(hearit.getFileUrl(fileType)).isEqualTo(fileUrl);
+    }
+}
```
아래 개행 하나 추가해야할 것 같아요!

gabean13 replied:
자꾸 EOF 안해서 죄송해욥
editorconfig 추가해서 아예 저장할때마다 eof 추가해주도록 했습니다

JO-eusan reviewed (changes requested):
가콩! 관리자 페이지에 데이터 수정 기능 확인했습니다.
프로덕션 코드에서 Validator에 대해 리뷰 남겨서 확인 부탁드립니다.😀

Byesol commented on backend/src/main/java/com/onair/hearit/domain/FileType.java:
```diff
@@ -0,0 +1,61 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum FileType {
```
현재 `FileType` enum이 `Hearit` 객체를 직접 인자로 받아서 `Hearit`의 내부 상태(URL 필드)를 변경하거나 조회하는 책임을 가지고 있고 Hearit에서도 결국 FileType을 인자로 받는 메서드가 있는 상태인데 이 부분이 이상하게 느껴집니당.

이렇게 되면 url들은  로직을 가지고 있는 하나의 값 객체이고, 따라서 Hearit Entity가 url들로 이루어진 하나의 값 객체를 들고 있어야 한다는 생각이 들어요.

Byesol replied:
이 부분은 낼 한 번 논의해보는 게 좋을 거 같네요.

Byesol commented on backend/src/main/java/com/onair/hearit/admin/application/FileStorageService.java:
```diff
@@ -44,4 +44,15 @@ private String uploadToS3(MultipartFile multipartFile, String key) {
+            log.error("S3 삭제 실패: " + key, e);
+            throw new S3Exception("S3 삭제 실패", e);
+        }
+    }
```
이거 고쳐야 할 것 같아요

Byesol commented on backend/src/test/java/com/onair/hearit/domain/HearitTest.java:
```diff
@@ -0,0 +1,117 @@
+
+        @ParameterizedTest
+        @DisplayName("제목은 35자 이하의 문자열이야한다.")
+        @ValueSource(strings = {"", "012345678901234567890123456789123456"})
```
와오 테스트 꼼꼼하게 해주셨네용.  좋네용

Byesol reviewed (commented):
가콩 고생 많으셨어요!

gabean13 committed:
[BE] chore: editorConfig 추가 #246

gabean13 committed:
[BE] rename: S3Exception -> AdminFileException으로 변경 #246

gabean13 committed:
[BE] style: 코드 컨벤션에 맞춰서 수정 #246

gabean13 committed:
[BE] Merge branch 'develop-be' into refactor/#246-admin-s3-update #246

# Conflicts:
#	backend/src/main/java/com/onair/hearit/auth/config/ApiSecurityConfig.java

gabean13 committed:
[BE] refactor: 삼항연산자 제거 #246

JO-eusan reviewed (approved):
리뷰 반영 확인했습니다. 추후 refactor Issue에서 만나엽 😎

Byesol reviewed (approved):
확인했습니다!
